### PR TITLE
CLN GH23123 Move SparseArray to arrays

### LINKED ISSUE #23123: Move SparseArray implementation to pandas/core/arrays
Just for organization / consistency, we should develop it next to our other arrays.

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #23123 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Importing `SparseArray` from the `Arrays` folder was causing circular import paths with the `__init__.py`, so I had to empty it and fix every other file that imported from `pandas.core.arrays`

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
This change in imports of other arrays everywhere should not be needed. Can you detail where the circular import problem was coming from?

**Comment by datapythonista** [comment]
@JustinZhengBC you've got some `print()` calls in your PR that I guess were for your own debugging. Can you remove them?

File: pandas/core/arrays/__init__.py [hunk 1]
```diff
@@ -1,10 +0,0 @@
-from .base import (ExtensionArray,    # noqa
-                   ExtensionOpsMixin,
-                   ExtensionScalarOpsMixin)
-from .categorical import Categorical  # noqa
-from .datetimes import DatetimeArrayMixin  # noqa
-from .interval import IntervalArray  # noqa
```
Comment (jreback) [inline pandas/core/arrays/__init__.py]: this is a nice pattern, pls restore this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by JustinZhengBC** [comment]
The main import loop problem is in `pandas.core.types.dtypes.common.py`, which imports `SparseDtype` from `pandas.core.arrays.sparse.dtype` for two uses in checking `is_instance`. Whenever a SparseArray is imported, it now executes `pandas.core.arrays.__init__.py`. Many of the files mentioned in `__init__.py` in turn import `common.py`, causing another import from SparseArray, causing another call to `__init__.py` in the process.

I will try putting the offending import statements from `common.py` into the functions where they are used so that they are not automatically called on import, and see if that gets rid of the loop.

**Comment by JustinZhengBC** [comment]
Almost all of the tests seem to be failing in the same place. The trace is below, and the offending line is `from pandas.core.arrays.sparse.array import _maybe_to_sparse`, where the tests claim they can't find `pandas.core.arrays.sparse.array`. The weird thing is, the tests do not fail locally, and I can even execute the exact same line without any errors, so I'm a bit stuck here. Am I doing something wrong, or is it that moving a folder and thus changing the API causes tests to fail because of how they're set up?

```
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/pandas/lib/python2.7/site-packages/_pytest/config/__init__.py", line 419, in _importconftest
    mod = conftestpath.pyimport()
  File "/home/travis/miniconda3/envs/pandas/lib/python2.7/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/home/travis/build/pandas-dev/pandas/pandas/__init__.py", line 42, in <module>
    from pandas.core.api import *
  File "/home/travis/build/pandas-dev/pandas/pandas/core/api.py", line 10, in <module>
    from pandas.core.groupby import Grouper
  File "/home/travis/build/pandas-dev/pandas/pandas/core/groupby/__init__.py", line 1, in <module>
    from pandas.core.groupby.groupby import GroupBy  # flake8: noqa
  File "/home/travis/build/pandas-dev/pandas/pandas/core/groupby/groupby.py", line 39, in <module>
    from pandas.core.generic import NDFrame
  File "/home/travis/build/pandas-dev/pandas/pandas/core/generic.py", line 43, in <module>
    from pandas.core.internals import BlockManager
  File "/home/travis/build/pandas-dev/pandas/pandas/core/internals/__init__.py", line 10, in <module>
    from .managers import (  # noqa:F401
  File "/home/travis/build/pandas-dev/pandas/pandas/core/internals/managers.py", line 32, in <module>
    from pandas.core.arrays.sparse.array import _maybe_to_sparse
ImportError: No module named sparse.array
ERROR: could not load /home/travis/build/pandas-dev/pandas/pandas/conftest.py
```

**Comment by datapythonista** [comment]
Did you try to simply `import pandas`? I think there is a wrong import in an `__init__.py`, or something similar.

**Comment by jorisvandenbossche** [comment]
@JustinZhengBC It seems you have removed the sparse tests, without adding them back in the `tests/arrays` folder.

Something general, should we keep the SparseSeries / SparseDataFrame where they were in `pandas.core.sparse` ? Since they are not really arrays, and we might deprecate them anyhow. Then moving them is maybe not that worth it?

**Comment by JustinZhengBC** [comment]
@jorisvandenbossche thanks for the catch. All the tests should be there now. The "files changed" tab still says I deleted `pandas/core/sparse/api.py` even though it also says I created another file with the same name in the right location. Other than that, it shouldn't say I deleted any other files now.

File: pandas/compat/numpy/function.py [hunk 2]
```diff
@@ -19,7 +19,8 @@
 """
 
 from numpy import ndarray
-from pandas.util._validators import (validate_args, validate_kwargs,
+from pandas.util._validators import (validate_args,
```
Comment (jreback) [inline pandas/compat/numpy/function.py]: is there a reason you are changing unrelated things? (even formatting)

**Review by jreback — CHANGES_REQUESTED** [review]
yeah I would *only* more the actual sparse array itself here, pls leave the SparseSeries/DataFrame exactly where they were (and there tests).

**Comment by TomAugspurger** [comment]
Test failure is unrelated to the changes here. Apparently, the SparseArray.unique implementation is flaky. Taking a look now.

**Comment by TomAugspurger** [comment]
```python

In [4]: pd.SparseArray([0, 0]).unique()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-c1befb06646c> in <module>
----> 1 pd.SparseArray([0, 0]).unique()

~/sandbox/pandas/pandas/core/sparse/array.py in unique(self)
    570     def unique(self):
    571         uniques = list(pd.unique(self.sp_values))
--> 572         fill_loc = self._first_fill_value_loc()
    573         if fill_loc >= 0:
    574             uniques.insert(fill_loc, self.fill_value)

~/sandbox/pandas/pandas/core/sparse/array.py in _first_fill_value_loc(self)
    562
    563         indices = self.sp_index.to_int_index().indices
--> 564         if indices[0] > 0:
    565             return 0
    566

IndexError: index 0 is out of bounds for axis 0 with size 0

```

**Comment by TomAugspurger** [comment]
And a fix

```diff
diff --git a/pandas/core/sparse/array.py b/pandas/core/sparse/array.py
index 15b5118db..17e2bd188 100644
--- a/pandas/core/sparse/array.py
+++ b/pandas/core/sparse/array.py
@@ -561,7 +561,7 @@ class SparseArray(PandasObject, ExtensionArray, ExtensionOpsMixin):
             return -1
 
         indices = self.sp_index.to_int_index().indices
-        if indices[0] > 0:
+        if len(indices) == 0 or indices[0] > 0:
             return 0
 
         diff = indices[1:] - indices[:-1]
diff --git a/pandas/tests/sparse/test_array.py b/pandas/tests/sparse/test_array.py
index 0257d9962..5b1afdc7a 100644
--- a/pandas/tests/sparse/test_array.py
+++ b/pandas/tests/sparse/test_array.py
@@ -1065,6 +1065,13 @@ def test_unique_na_fill(arr, fill_value):
     tm.assert_numpy_array_equal(a, b)
 
 
+def test_unique_all_sparse():
+    arr = SparseArray([0, 0])
+    result = arr.unique()
+    expected = SparseArray([0])
+    tm.assert_sp_array_equal(result, expected)
+
+
 def test_map():
     arr = SparseArray([0, 1, 2])
     expected = SparseArray([10, 11, 12], fill_value=10)
```

@jreback @jorisvandenbossche any objections to including that fix here? It seems simple enough to not need another PR, which will create a merge conflict here. Or we can merge this and do that in a separate PR.

**Comment by TomAugspurger** [comment]
Found another random failure. Going to split these two out to a new PR.

**Comment by TomAugspurger** [comment]
Updates:

1. Moved pandas/core/arrays/sparse/array.py to pandas/core/arrays/sparse.py
2. Moved SparseDtype implementation into `pandas/core/arrays/sparse.py`
3. Moved `sparse/{series,frame,api,scipy_sparse}.py ` back to `pandas/core/sparse`.

File: pandas/core/arrays/datetimes.py [hunk 3]
```diff
@@ -523,7 +523,7 @@ def _add_delta(self, delta):
         The result's name is set outside of _add_delta by the calling
         method (__add__ or __sub__)
         """
-        from pandas.core.arrays.timedeltas import TimedeltaArrayMixin
+        from pandas.core.arrays import TimedeltaArrayMixin
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimes.py]: Were these changes here before mine? I may revert them.

**Comment by TomAugspurger** [comment]
I reverted some extraneous changes in fe53b50. @JustinZhengBC if you're interested, you can re-revert that commit later and make a separate PR with those sytlisitc changes. I notice now that it has a formatting change too in https://github.com/pandas-dev/pandas/pull/23147/files#diff-6cdf34ac065f3e1e350f94c70e76898bL6, sorry. You'll need to manually fix that if you re-revert it.

File: pandas/api/extensions/__init__.py [hunk 4]
```diff
@@ -3,8 +3,8 @@
                                   register_index_accessor,
                                   register_series_accessor)
 from pandas.core.algorithms import take  # noqa
-from pandas.core.arrays.base import (ExtensionArray,    # noqa
-                                     ExtensionScalarOpsMixin)
+from pandas.core.arrays import (ExtensionArray,    # noqa
```
Comment (TomAugspurger) [inline pandas/api/extensions/__init__.py]: Ah, should have reverted this too :/ Oh well.

**Comment by TomAugspurger** [comment]
@jreback @jorisvandenbossche @datapythonista could I get a quick sanity check on this? I'd like to merge on green since I'll have some followup PRs that will be blocked by this.

**Comment by jreback** [comment]
lgtm at a glance

**Comment by jorisvandenbossche** [comment]
Looks good to me as well based on a quick view

**Comment by TomAugspurger** [comment]
Merging so we can move forward. Will keep an eye on master though.

**Comment by TomAugspurger** [comment]
Thanks @JustinZhengBC!

**Pull request merged**